### PR TITLE
Adressing libf3d API Public Review

### PR DESCRIPTION
**Pull request by mwestphal**
- adding recursive to addFIles
- improving options doc
- fix setIcon type
- better camera doc and typing
- rework exceptions
- missing include
- using includes instead of fwd decl when it makes sense
- rule of 5

see https://github.com/f3d-app/f3d/pull/351

### CODE REVIEW

**Comment by mwestphal** [comment]
@Connor-Bowley ptal :)

File: library/public/engine.h [hunk 1]
```diff
@@ -21,38 +21,31 @@
 #ifndef f3d_engine_h
 #define f3d_engine_h
 
+#include "exception.h"
 #include "export.h"
```
Comment (Meakk) [inline library/public/engine.h:25]: I'm honestly not sure about this change. There are some drawbacks doing this, like increasing the compilation time for the user. I still think forward decl is better.

Comment (Meakk) [reply library/public/engine.h:25]: If we want to simplify usability, we can provide a f3d.h file which include all other files. This way, users who want to include only required headers can still do it.

**Review by Meakk — CHANGES_REQUESTED** [review]

File: library/public/camera.h [hunk 2]
```diff
@@ -24,25 +25,39 @@ namespace f3d
 class F3D_EXPORT camera
 {
 public:
-  typedef std::array<double, 3> vector3_t;
-  typedef std::array<double, 16> matrix4_t;
+  //@{
+  /**
+   * Camera specific types.
+   */
+  struct point3_t : std::array<double, 3>
```
[the remark this answers is not in the document]
Comment (Connor-Bowley) [reply library/public/camera.h:32]: This is more along the lines of strong typing, which will enable the compiler to prevent doing things like accidentally using a vector when meaning to use a point. There is more type safety.

[the remark this answers is not in the document]
Comment (mwestphal) [reply library/public/camera.h:32]: Yup, that would be useless without the struct/class

File: library/public/engine.h (hunk 1 above)
Comment (mwestphal) [reply library/public/engine.h:25]: With less than 10 headers, I think there is no increase. Did you read the link provided by @Connor-Bowley ?

File: library/public/camera.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meakk) [reply library/public/camera.h:32]: Ok, then it should probably goes in its own file, or at least move it outside of the camera class scope, so we can reuse these types in other places later.

File: library/public/engine.h (hunk 1 above)
Comment (Meakk) [reply library/public/engine.h:25]: Fair enough. I'm not used to that but I guess it's an old habit

File: library/public/camera.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwestphal) [reply library/public/camera.h:32]: good point

[the remark this answers is not in the document]
Comment (Connor-Bowley) [reply library/public/camera.h:32]: Agree with its own file outside of camera. I would think you would also want to add some math operators and things like `operator<<`.

[the remark this answers is not in the document]
Comment (mwestphal) [reply library/public/camera.h:32]: it may be added later, this feels a bit outside of the scope of the first version imo.

File: library/CMakeLists.txt [hunk 3]
```diff
@@ -68,12 +68,12 @@ configure_file(
 
 # Define libf3d target
 set(F3D_SOURCE_FILES
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/camera_impl.cxx
   ${CMAKE_CURRENT_BINARY_DIR}/src/config.cxx
   ${CMAKE_CURRENT_SOURCE_DIR}/src/animationManager.cxx
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/init.cxx
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/image.cxx
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/camera_impl.cxx
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply library/CMakeLists.txt:73]: they are sorted, but some are using binary dir. ill fix them manually

File: library/public/types.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * Describe a 4x4 matrix
+ */
+struct matrix4_t : std::array<double, 16>
+{
+};
+
+/**
+ * Describe an angle in degrees
+ */
+typedef double angleDeg_t;
```
Comment (Meakk) [inline library/public/types.h:37]:
```suggestion
using angleDeg_t = double;
```

File: library/src/camera_impl.cxx [hunk 5]
```diff
@@ -33,22 +33,22 @@ camera& camera_impl::setPosition(const vector3_t& pos)
 }
 
 //----------------------------------------------------------------------------
-camera::vector3_t camera_impl::getPosition()
+f3d::point3_t camera_impl::getPosition()
```
Comment (Meakk) [inline library/src/camera_impl.cxx:36]:
```suggestion
point3_t camera_impl::getPosition()
```

File: library/src/camera_impl.cxx [hunk 6]
```diff
@@ -58,15 +58,15 @@ camera& camera_impl::setFocalPoint(const vector3_t& foc)
 }
 
 //----------------------------------------------------------------------------
-camera::vector3_t camera_impl::getFocalPoint()
+f3d::point3_t camera_impl::getFocalPoint()
```
Comment (Meakk) [inline library/src/camera_impl.cxx:61]:
```suggestion
point3_t camera_impl::getFocalPoint()
```

File: library/src/camera_impl.cxx [hunk 7]
```diff
@@ -83,7 +83,7 @@ camera& camera_impl::setViewUp(const vector3_t& up)
 }
 
 //----------------------------------------------------------------------------
-camera::vector3_t camera_impl::getViewUp()
+f3d::vector3_t camera_impl::getViewUp()
```
Comment (Meakk) [inline library/src/camera_impl.cxx:86]:
```suggestion
vector3_t camera_impl::getViewUp()
```

File: library/src/camera_impl.cxx [hunk 8]
```diff
@@ -108,15 +108,15 @@ camera& camera_impl::setViewAngle(const double& angle)
 }
 
 //----------------------------------------------------------------------------
-double camera_impl::getViewAngle()
+f3d::angleDeg_t camera_impl::getViewAngle()
```
Comment (Meakk) [inline library/src/camera_impl.cxx:111]:
```suggestion
angleDeg_t camera_impl::getViewAngle()
```

File: library/src/camera_impl.cxx [hunk 9]
```diff
@@ -134,7 +134,7 @@ camera& camera_impl::setViewMatrix(const matrix4_t& matrix)
 }
 
 //----------------------------------------------------------------------------
-camera::matrix4_t camera_impl::getViewMatrix()
+f3d::matrix4_t camera_impl::getViewMatrix()
```
Comment (Meakk) [inline library/src/camera_impl.cxx:137]:
```suggestion
matrix4_t camera_impl::getViewMatrix()
```

File: library/src/image.cxx [hunk 10]
```diff
@@ -25,19 +25,13 @@ class image::internals
 
   vtkSmartPointer<vtkImageImport> GetVTKImporter() const
   {
-    if (this->Buffer.size() != this->Width * this->Height * this->Channels)
```
Comment (Meakk) [inline library/src/image.cxx:28]: Why did you remove this block?

File: python/F3DPythonBindings.cxx [hunk 11]
```diff
@@ -92,7 +92,7 @@ PYBIND11_MODULE(f3d, module)
     .export_values();
 
   loader
-    .def("addFiles", &f3d::loader::addFiles, "Add files to the loader", py::arg("list of files"))
+    .def("addFiles", &f3d::loader::addFiles, "Add files to the loader", py::arg("list of files"), py::arg("recursive") = false)
```
Comment (Meakk) [inline python/F3DPythonBindings.cxx:95]: This file will need more changes, but it can be done later

**Review by Meakk — CHANGES_REQUESTED** [review]

File: python/F3DPythonBindings.cxx (hunk 11 above)
Comment (mwestphal) [reply python/F3DPythonBindings.cxx:95]: yes

File: library/src/image.cxx (hunk 10 above)
Comment (mwestphal) [reply library/src/image.cxx:28]: Because this test just cannot be true. I could put an assert if you want.

Comment (Meakk) [reply library/src/image.cxx:28]: Yes please

File: library/private/camera_impl.h [hunk 12]
```diff
@@ -35,28 +35,28 @@ class camera_impl : public camera
...
+  camera& setFocalPoint(const point3_t& foc) override;
+  point3_t getFocalPoint() override;
+  void getFocalPoint(point3_t& foc) override;
   camera& setViewUp(const vector3_t& up) override;
   vector3_t getViewUp() override;
   void getViewUp(vector3_t& up) override;
-  camera& setViewAngle(const double& angle) override;
-  double getViewAngle() override;
-  void getViewAngle(double& angle) override;
+  camera& setViewAngle(const angleDeg_t& angle) override;
```
Comment (Meakk) [inline library/private/camera_impl.h:47]: I think it should be snake case: `angle_deg_t`. See https://en.cppreference.com/w/cpp/types/max_align_t for example

File: library/src/camera_impl.cxx [hunk 13]
```diff
@@ -33,22 +33,22 @@ camera& camera_impl::setPosition(const vector3_t& pos)
...
-void camera_impl::getPosition(vector3_t& pos)
+void camera_impl::getPosition(point3_t& pos)
 {
   vtkCamera* cam = this->GetVTKCamera();
   cam->GetPosition(pos.data());
 }
 
 //----------------------------------------------------------------------------
-camera& camera_impl::setFocalPoint(const vector3_t& foc)
+camera& camera_impl::setFocalPoint(const point3_t& foc)
```
Comment (Meakk) [inline library/src/camera_impl.cxx:51]: Can you use initializer list here with the current implementation? I'm expecting user wanting to call `setFocalPoint({ 0, 0, 0})`

**Review by Meakk — CHANGES_REQUESTED** [review]

File: library/src/camera_impl.cxx (hunk 13 above)
Comment (mwestphal) [reply library/src/camera_impl.cxx:51]: should I put that in types.h or in camera_impl::setFocalPoint ?

Comment (Meakk) [reply library/src/camera_impl.cxx:51]: Actually, Maybe just add setters with 3 doubles.

Comment (Connor-Bowley) [reply library/src/camera_impl.cxx:51]: I do think that making sure that `std::array<double, 3>`'s initializer list constructor is visible would be useful. The downside of setters with 3 doubles is the loss of typing.

Comment (mwestphal) [reply library/src/camera_impl.cxx:51]: > Actually, Maybe just add setters with 3 doubles.

I dont think this is the right way to go, why even defining a vector3_t in that case ?

> that std::array<double, 3>'s initializer list constructor is visible would be useful. 

I also think so

File: library/private/camera_impl.h (hunk 12 above)
Comment (mwestphal) [reply library/private/camera_impl.h:47]: why should typedef style be different than local variables ? Afaik, cppref seems to use snake case everywhere.

Comment (Meakk) [reply library/private/camera_impl.h:47]: If you choose to name the types with the _t suffix, it's to mimic the STL types, so we should use snake case in this case. I don't like point_3_t though, I think point3_t looks fine.

**Review by Meakk — APPROVED** [review]
